File: server/chat_app_server/app/models/session.py
#!/usr/bin/env python3
"""
会话相关的数据模型
"""
from pydantic import BaseModel
from typing import Optional, List, Dict, Any
from datetime import datetime
from .database_factory import get_database
import asyncio
import uuid


class SessionCreate(BaseModel):
    """会话创建模型"""
    title: str
    description: Optional[str] = None
    metadata: Optional[Dict[str, Any]] = None


class SessionMcpServerCreate(BaseModel):
    """会话MCP服务器关联创建模型"""
    session_id: str
    mcp_server_name: str
    config: Optional[Dict[str, Any]] = None


class SessionService:
    """会话数据服务"""
    
    @classmethod
    def create(cls, session_data: SessionCreate) -> str:
        """创建会话"""
        session_id = str(uuid.uuid4())
        
        db = get_database()
        db.execute_sync(
            """
            INSERT INTO sessions (id, title, description, metadata, created_at, updated_at)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (session_id, session_data.title, session_data.description, 
             str(session_data.metadata) if session_data.metadata else None, 
             datetime.now(), datetime.now())
        )
        return session_id
    
    @classmethod
    def get_all(cls) -> List[Dict[str, Any]]:
        """获取所有会话"""
        db = get_database()
        rows = db.fetchall_sync("SELECT * FROM sessions ORDER BY updated_at DESC")
        return [dict(row) for row in rows]
    
    @classmethod
    def get_by_id(cls, session_id: str) -> Optional[Dict[str, Any]]:
        """根据ID获取会话"""
        db = get_database()
        row = db.fetchone_sync("SELECT * FROM sessions WHERE id = ?", (session_id,))
        return dict(row) if row else None
    
    @classmethod
    def delete(cls, session_id: str) -> bool:
        """删除会话"""
        db = get_database()
        cursor = db.execute_sync("DELETE FROM sessions WHERE id = ?", (session_id,))
        return cursor.rowcount > 0
    
    @classmethod
    async def delete_async(cls, session_id: str) -> bool:
        """异步删除会话"""
        return await asyncio.to_thread(cls.delete, session_id)


class SessionMcpServerService:
    """会话MCP服务器关联数据服务"""
    
    @classmethod
    def create(cls, server_data: SessionMcpServerCreate) -> int:
        """创建会话MCP服务器关联"""
        db = get_database()
        cursor = db.execute_sync(
            """
            INSERT INTO session_mcp_servers (session_id, mcp_server_name, config, created_at)
            VALUES (?, ?, ?, ?)
            """,
            (server_data.session_id, server_data.mcp_server_name, 
             str(server_data.config) if server_data.config else None, datetime.now())
        )
        return cursor.lastrowid
    
    @classmethod
    def get_by_session(cls, session_id: str) -> List[Dict[str, Any]]:
        """根据会话ID获取MCP服务器关联"""
        db = get_database()
        rows = db.fetchall_sync("SELECT * FROM session_mcp_servers WHERE session_id = ?", (session_id,))
        return [dict(row) for row in rows]
    
    @classmethod
    def delete_by_session(cls, session_id: str) -> bool:
        """根据会话ID删除MCP服务器关联"""
        db = get_database()
        cursor = db.execute_sync("DELETE FROM session_mcp_servers WHERE session_id = ?", (session_id,))
        return cursor.rowcount > 0
    
    @classmethod
    async def delete_by_session_async(cls, session_id: str) -> bool:
        """异步根据会话ID删除MCP服务器关联"""
        return await asyncio.to_thread(cls.delete_by_session, session_id)


# 为了向后兼容，保留旧的类名作为别名
# 这样现有的代码仍然可以工作
class SessionOperations:
    """向后兼容的会话操作类（已弃用，请使用 SessionService）"""
    
    @classmethod
    def create(cls, session_data):
        """创建会话 - 向后兼容方法"""
        if isinstance(session_data, dict):
            # 如果传入的是字典，转换为 SessionCreate 模型
            session_model = SessionCreate(**session_data)
            return SessionService.create(session_model)
        elif hasattr(session_data, 'title'):
            # 如果传入的是 SessionCreate 模型
            return SessionService.create(session_data)
        else:
            # 如果是旧的调用方式（直接传参数）
            return SessionService.create(SessionCreate(
                title=session_data,
                description=None,
                metadata=None
            ))
    
    @classmethod
    def get_all(cls):
        return SessionService.get_all()
    
    @classmethod
    def get_by_id(cls, session_id: str):
        return SessionService.get_by_id(session_id)
    
    @classmethod
    def delete(cls, session_id: str):
        return SessionService.delete(session_id)
    
    @classmethod
    async def delete_async(cls, session_id: str):
        return await SessionService.delete_async(session_id)

File: server/chat_app_node_server/src/models/session.js
/**
 * Session 数据模型 - 会话管理
 * 复刻自 Python: app/models/session.py
 */

import { v4 as uuidv4 } from 'uuid';
import { getDatabaseSync } from './database-factory.js';

/**
 * Session 模型类
 */
export class Session {
  constructor(data) {
    this.id = data.id || uuidv4();
    this.title = data.title;
    this.description = data.description || null;
    this.metadata = data.metadata || null;
    this.user_id = data.user_id || null;
    this.project_id = data.project_id || null;
    this.created_at = data.created_at || new Date().toISOString();
    this.updated_at = data.updated_at || new Date().toISOString();
  }

  /**
   * 转换为数据库格式
   */
  toDb() {
    return {
      id: this.id,
      title: this.title,
      description: this.description,
      metadata: this.metadata ? JSON.stringify(this.metadata) : null,
      user_id: this.user_id,
      project_id: this.project_id,
      created_at: this.created_at,
      updated_at: this.updated_at
    };
  }

  /**
   * 从数据库行创建实例
   */
  static fromDb(row) {
    if (!row) return null;
    return new Session({
      ...row,
      metadata: row.metadata ? JSON.parse(row.metadata) : null
    });
  }
}

/**
 * Session 服务类
 */
export class SessionService {
  /**
   * 创建会话（异步）
   */
  static async create(sessionData) {
    const session = new Session(sessionData);
    const db = getDatabaseSync();
    const dbData = session.toDb();

    await db.execute(
      `INSERT INTO sessions (id, title, description, metadata, user_id, project_id, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        dbData.id,
        dbData.title,
        dbData.description,
        dbData.metadata,
        dbData.user_id,
        dbData.project_id,
        dbData.created_at,
        dbData.updated_at
      ]
    );

    return session.id;
  }

  /**
   * 创建会话（同步）
   */
  static createSync(sessionData) {
    const session = new Session(sessionData);
    const db = getDatabaseSync();
    const dbData = session.toDb();

    db.executeSync(
      `INSERT INTO sessions (id, title, description, metadata, user_id, project_id, created_at, updated_at)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?)`,
      [
        dbData.id,
        dbData.title,
        dbData.description,
        dbData.metadata,
        dbData.user_id,
        dbData.project_id,
        dbData.created_at,
        dbData.updated_at
      ]
    );

    return session.id;
  }

  /**
   * 根据 ID 获取会话（异步）
   */
  static async getById(sessionId) {
    const db = getDatabaseSync();
    const row = await db.fetchone('SELECT * FROM sessions WHERE id = ?', [sessionId]);
    return Session.fromDb(row);
  }

  /**
   * 获取所有会话（异步）
   */
  static async getAll() {
    const db = getDatabaseSync();
    const rows = await db.fetchall('SELECT * FROM sessions ORDER BY created_at DESC');
    return rows.map(row => Session.fromDb(row));
  }

  /**
   * 根据用户和项目获取会话（异步）
   */
  static async getByUserProject(userId = null, projectId = null) {
    const db = getDatabaseSync();
    let query = 'SELECT * FROM sessions WHERE 1=1';
    const params = [];

    if (userId) {
      query += ' AND user_id = ?';
      params.push(userId);
    }

    if (projectId) {
      query += ' AND project_id = ?';
      params.push(projectId);
    }

    query += ' ORDER BY created_at DESC';

    const rows = await db.fetchall(query, params);
    return rows.map(row => Session.fromDb(row));
  }

  /**
   * 删除会话（异步）
   */
  static async delete(sessionId) {
    const db = getDatabaseSync();
    await db.execute('DELETE FROM sessions WHERE id = ?', [sessionId]);
  }

  /**
   * 更新会话（异步）
   */
  static async update(sessionId, updates) {
    const db = getDatabaseSync();
    const setClause = [];
    const params = [];

    if (updates.title !== undefined) {
      setClause.push('title = ?');
      params.push(updates.title);
    }

    if (updates.description !== undefined) {
      setClause.push('description = ?');
      params.push(updates.description);
    }

    if (updates.metadata !== undefined) {
      setClause.push('metadata = ?');
      params.push(JSON.stringify(updates.metadata));
    }

    setClause.push('updated_at = ?');
    params.push(new Date().toISOString());

    params.push(sessionId);

    const query = `UPDATE sessions SET ${setClause.join(', ')} WHERE id = ?`;
    await db.execute(query, params);
  }
}
